File: saral_hr/saral_hr/report/retention_deposit_register/retention_deposit_register.js
// Copyright (c) 2026, sj and contributors
// For license information, please see license.txt

frappe.query_reports["Retention Deposit Register"] = {
    filters: [
        {
            fieldname: "year",
            label: __("Year"),
            fieldtype: "Select",
            options: get_year_options(),
            reqd: 1,
            default: ""
        },
        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: [
                "", "January", "February", "March", "April",
                "May", "June", "July", "August",
                "September", "October", "November", "December"
            ],
            reqd: 1,
            default: ""
        },
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "MultiSelectList",
            reqd: 1,
            get_data: function(txt) {
                return frappe.db.get_link_options("Company", txt);
            }
        },
        {
            fieldname: "category",
            label: __("Category"),
            fieldtype: "Link",
            options: "Category",
            reqd: 1,
            default: ""
        },
        {
            fieldname: "employee",
            label: __("Employee"),
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                let year      = frappe.query_report.get_filter_value("year");
                let month     = frappe.query_report.get_filter_value("month");
                let companies = frappe.query_report.get_filter_value("company");
                let category  = frappe.query_report.get_filter_value("category");

                return frappe.call({
                    method: "saral_hr.saral_hr.report.retention_deposit_register.retention_deposit_register.get_retention_employees_for_filter",
                    args: {
                        year:      year,
                        month:     month,
                        companies: JSON.stringify(companies || []),
                        category:  category || "",
                        txt:       txt || ""
                    }
                }).then(r => {
                    return (r.message || []).map(emp => ({
                        value:       emp.employee,
                        description: emp.employee_name || ""
                    }));
                });
            }
        }
    ],

    onload: function(report) {
        frappe.after_ajax(() => {
            report.page.set_primary_action(__("Print"), () => {
                retention_custom_print(report);
            }, "printer");
        });
    }
};

// ─────────────────────────────────────────────────────────────────────────────

function get_year_options() {
    let current_year = new Date().getFullYear();
    let years = [""];
    for (let i = current_year - 2; i <= current_year + 1; i++) {
        years.push(i.toString());
    }
    return years;
}

function retention_custom_print(report) {
    let data    = frappe.query_report.data || [];
    let filters = frappe.query_report.get_values() || {};

    let month = filters.month || "";
    let year  = filters.year  || "";

    let rows       = data.filter(r => r.employee_name !== "Total");
    let totals_row = data.find(r => r.employee_name === "Total") || {};

    let tbody = rows.map((row, idx) => `
        <tr>
            <td class="center">${idx + 1}</td>
            <td>${row.employee_id || ""}</td>
            <td>${row.employee_name || ""}</td>
            <td class="center">${fmt_date(row.date_of_joining)}</td>
            <td class="center">${fmt_date(row.ded_upto)}</td>
            <td class="num">${fmt(row.retention_amount)}</td>
        </tr>
    `).join("");

    tbody += `
        <tr class="total-row">
            <td colspan="5" style="text-align:right; padding-right:12px; font-weight:bold;">Total</td>
            <td class="num">${fmt(totals_row.retention_amount)}</td>
        </tr>
    `;

    let html = `
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <title>Retention Deposit Register – ${month} ${year}</title>
            <style>
                * { margin:0; padding:0; box-sizing:border-box; }
                body { font-family: Arial, sans-serif; font-size: 11px; color: #000; }
                .container { padding: 24px; }

                .header { text-align: center; margin-bottom: 16px; }
                .header .company-name { font-size: 16px; font-weight: bold; text-transform: uppercase; }
                .header .report-title { font-size: 13px; font-weight: bold; margin-top: 4px; }
                .header .period       { font-size: 11px; margin-top: 2px; color: #444; }

                table { width: 100%; border-collapse: collapse; margin-top: 12px; }

                thead tr th {
                    background-color: #e8e8e8;
                    border: 1px solid #999;
                    padding: 7px 8px;
                    text-align: center;
                    font-size: 11px;
                    font-weight: bold;
                }

                tbody tr td {
                    border: 1px solid #bbb;
                    padding: 6px 8px;
                    font-size: 11px;
                }

                tbody tr:nth-child(even):not(.total-row) { background-color: #f9f9f9; }

                td.num    { text-align: right; }
                td.center { text-align: center; }

                tr.total-row td {
                    font-weight: bold;
                    background-color: #eef4fb;
                    border-top: 2px solid #555;
                    border-bottom: 2px solid #555;
                }

                .footer {
                    margin-top: 40px;
                    display: flex;
                    justify-content: space-between;
                }
                .sign-block { text-align: center; width: 180px; }
                .sign-block .line  { border-top: 1px solid #000; margin-bottom: 4px; }
                .sign-block .label { font-size: 10px; color: #444; }

                @media print {
                    @page { size: A4 landscape; margin: 15mm; }
                    body  { -webkit-print-color-adjust: exact; print-color-adjust: exact; }
                }
            </style>
        </head>
        <body>
        <div class="container">

            <div class="header">
                <div class="company-name">${frappe.boot.sysdefaults.company || ""}</div>
                <div class="report-title">Retention Deposit Register</div>
                <div class="period">For the Month of ${month} ${year}</div>
            </div>

            <table>
                <thead>
                    <tr>
                        <th style="width:55px;">Sr. No.</th>
                        <th style="width:140px;">Employee ID</th>
                        <th style="width:200px;">Employee Name</th>
                        <th style="width:140px;">Date of Joining</th>
                        <th style="width:160px;">Ded. Upto (3 Yrs)</th>
                        <th style="width:150px;">Retention Deposit</th>
                    </tr>
                </thead>
                <tbody>
                    ${tbody}
                </tbody>
            </table>

            <div class="footer">
                <div class="sign-block">
                    <div class="line"></div>
                    <div class="label">Prepared By</div>
                </div>
                <div class="sign-block">
                    <div class="line"></div>
                    <div class="label">Checked By</div>
                </div>
                <div class="sign-block">
                    <div class="line"></div>
                    <div class="label">Authorised Signatory</div>
                </div>
            </div>

        </div>
        </body>
        </html>
    `;

    let w = window.open("", "_blank");
    w.document.write(html);
    w.document.close();
    w.focus();
    setTimeout(() => { w.print(); }, 500);
}

function fmt(val) {
    return parseFloat(val || 0).toFixed(2);
}

function fmt_date(val) {
    if (!val) return "-";
    let parts = val.split ? val.split("-") : [];
    if (parts.length === 3) {
        return `${parts[2]}-${parts[1]}-${parts[0]}`;
    }
    return val;
}

File: saral_hr/saral_hr/report/retention_deposit_register/retention_deposit_register.py
# Copyright (c) 2026, sj and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt, getdate, add_months

COMPONENT_NAME = "Retention"

MONTH_MAP = {
    'January': 1, 'February': 2, 'March': 3, 'April': 4,
    'May': 5, 'June': 6, 'July': 7, 'August': 8,
    'September': 9, 'October': 10, 'November': 11, 'December': 12
}


def execute(filters=None):
    filters = filters or {}
    validate_filters(filters)
    columns = get_columns()
    data    = get_data(filters)
    return columns, data


def validate_filters(filters):
    if not filters.get("month"):
        frappe.throw("Please select a Month")
    if not filters.get("year"):
        frappe.throw("Please select a Year")
    if not filters.get("company"):
        frappe.throw("Please select at least one Company")
    if not filters.get("category"):
        frappe.throw("Please select a Category")


def get_columns():
    return [
        {
            "label": "Sr. No.",
            "fieldname": "sr_no",
            "fieldtype": "Int",
            "width": 70
        },
        {
            "label": "Employee ID",
            "fieldname": "employee_id",
            "fieldtype": "Data",
            "width": 160
        },
        {
            "label": "Employee Name",
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 220
        },
        {
            "label": "Date of Joining",
            "fieldname": "date_of_joining",
            "fieldtype": "Date",
            "width": 150
        },
        {
            "label": "Deduction Upto (3 Years)",
            "fieldname": "ded_upto",
            "fieldtype": "Date",
            "width": 180
        },
        {
            "label": "Retention Deposit",
            "fieldname": "retention_amount",
            "fieldtype": "Float",
            "precision": 2,
            "width": 180
        },
    ]


def get_data(filters):
    import json

    month          = filters.get("month")
    year           = filters.get("year")
    category       = filters.get("category")       # single string (Link field)
    company_filter = filters.get("company")

    month_num  = MONTH_MAP.get(month)
    start_date = f"{year}-{month_num:02d}-01"

    conditions   = ""
    query_params = {"start_date": start_date, "component": COMPONENT_NAME}

    if company_filter:
        if isinstance(company_filter, str):
            company_filter = json.loads(company_filter)
        if company_filter:
            conditions += " AND ss.company IN %(companies)s"
            query_params["companies"] = tuple(company_filter)

    employee_filter = filters.get("employee")
    if employee_filter:
        if isinstance(employee_filter, str):
            employee_filter = json.loads(employee_filter)
        if employee_filter:
            conditions += " AND ss.employee IN %(employees)s"
            query_params["employees"] = tuple(employee_filter)

    # Category filter via Company Link (same pattern as all other reports)
    category_join = ""
    if category:
        category_join = """
            INNER JOIN `tabCompany Link` cl
                ON  cl.name     = ss.employee
                AND cl.category = %(category)s
        """
        query_params["category"] = category

    rows = frappe.db.sql("""
        SELECT
            ss.employee      AS employee_id,
            ss.employee_name AS employee_name,
            sd.amount        AS retention_amount
        FROM
            `tabSalary Slip` ss
        INNER JOIN
            `tabSalary Details` sd
            ON  sd.parent           = ss.name
            AND sd.parenttype       = 'Salary Slip'
            AND sd.parentfield      = 'deductions'
            AND sd.salary_component = %(component)s
            AND sd.amount > 0
        {category_join}
        WHERE
            ss.docstatus      = 1
            AND ss.start_date = %(start_date)s
            {conditions}
        ORDER BY
            ss.employee_name
    """.format(conditions=conditions, category_join=category_join), query_params, as_dict=1)

    if not rows:
        return []

    # Fetch date_of_joining from Company Link
    employee_ids = list({r.employee_id for r in rows})

    cl_rows = frappe.db.sql("""
        SELECT name, date_of_joining
        FROM `tabCompany Link`
        WHERE name IN %(ids)s
    """, {"ids": tuple(employee_ids)}, as_dict=1)

    cl_map = {r["name"]: r["date_of_joining"] for r in cl_rows}

    data            = []
    total_retention = 0.0

    for idx, row in enumerate(rows, start=1):
        doj      = cl_map.get(row.employee_id)
        ded_upto = None

        if doj:
            try:
                ded_upto = add_months(getdate(doj), 36)
            except Exception:
                ded_upto = None

        retention = flt(row.retention_amount, 2)
        total_retention += retention

        data.append({
            "sr_no":            idx,
            "employee_id":      row.employee_id,
            "employee_name":    row.employee_name,
            "date_of_joining":  doj,
            "ded_upto":         ded_upto,
            "retention_amount": retention,
        })

    # Totals row
    data.append({
        "sr_no":            "",
        "employee_id":      "",
        "employee_name":    "Total",
        "date_of_joining":  None,
        "ded_upto":         None,
        "retention_amount": flt(total_retention, 2),
        "bold":             1,
    })

    return data


@frappe.whitelist()
def get_retention_employees_for_filter(year=None, month=None, companies=None, category=None, txt=""):
    """
    Returns employees who have a Retention deduction in their submitted salary slip
    for the period, filtered by company and category.
    Used by the Employee MultiSelectList filter.
    """
    if not month or not year:
        return []

    month_num = MONTH_MAP.get(month)
    if not month_num:
        return []

    import json
    if isinstance(companies, str):
        try:
            companies = json.loads(companies)
        except Exception:
            companies = []

    start_date = f"{year}-{month_num:02d}-01"
    txt_filter = f"%{txt}%"

    company_condition = ""
    category_join     = ""
    params = {
        "component":  COMPONENT_NAME,
        "start_date": start_date,
        "txt":        txt_filter,
    }

    if companies:
        company_condition = "AND ss.company IN %(companies)s"
        params["companies"] = tuple(companies)

    if category:
        category_join = """
            INNER JOIN `tabCompany Link` cl
                ON  cl.name     = ss.employee
                AND cl.category = %(category)s
        """
        params["category"] = category

    results = frappe.db.sql("""
        SELECT DISTINCT
            ss.employee,
            ss.employee_name
        FROM
            `tabSalary Slip` ss
        INNER JOIN
            `tabSalary Details` sd
            ON  sd.parent           = ss.name
            AND sd.parenttype       = 'Salary Slip'
            AND sd.parentfield      = 'deductions'
            AND sd.salary_component = %(component)s
            AND sd.amount > 0
        {category_join}
        WHERE
            ss.docstatus      = 1
            AND ss.start_date = %(start_date)s
            AND (ss.employee LIKE %(txt)s OR ss.employee_name LIKE %(txt)s)
            {company_condition}
        ORDER BY ss.employee_name
        LIMIT 50
    """.format(category_join=category_join, company_condition=company_condition), params, as_dict=1)

    return results